src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs

```diff
@@ -728,11 +728,26 @@ private void EnsurePrivateImplementationDetailsStaticConstructor(PrivateImplemen
         /// </summary>
         private sealed class SynthesizedDefinitions
         {
-            public ConcurrentQueue<Cci.INestedTypeDefinition> NestedTypes;
+            private ConcurrentQueue<Cci.INestedTypeDefinition> NestedTypes;
             public ConcurrentQueue<Cci.IMethodDefinition> Methods;
             public ConcurrentQueue<Cci.IPropertyDefinition> Properties;
             public ConcurrentQueue<Cci.IFieldDefinition> Fields;
 
+            // Nested types may be queued from concurrent threads, but we need to emit them
+            // in a deterministic order.
+            internal IEnumerable<Cci.INestedTypeDefinition> DeterministicNestedTypes
+                => NestedTypes?.OrderBy(t => t.Name, StringComparer.Ordinal);
```

Comment:
what if you have multiple nested types with the same name, but different arity?

Reply:
why not emit them in source-order (taking into account partials as well of course).

Reply:
These nested types are synthesized. Not directly from source. I don't think we synthesize nested types with different arities. Could add this in 17.1 for extra safety (future-proofing)...
As for the order itself, I don't think there's a reason to prefer one specific order to another, as long as it's deterministic.

Reply:
> I don't think we synthesize nested types with different arities

Can you add a contract call to that effect hten?  if that assumption ever changes we want to update this.

Reply:
Good idea. I'll create a 17.1 PR to add assertion. Thanks

Reply:
Actually, since this PR isn't merged yet, I'll add the assertion here. Still have time for tomorrow's deadline